Fix crash during shutdown in distributed training

Flamefire opened the pull request:
Releasing and reaquiring the GIL during Python shutdown is not possible and leads to force termination.
Remove that.

Fixes #50853

Flamefire committed:
Fix crash during shutdown in distributed training

Releasing and reaquiring the GIL during Python shutdown is not possible and leads to force termination.
Remove that.

Fixes #50853

gbaned commented:
Hi @mdanatg Can you please review this PR ? Thank you!

mdanatg commented:
@gbaned Sorry for the delay. I no longer have approval rights on the tensorflow repo...

rainwoodman reviewed (changes requested):
Thanks for the fix. Could you add a comment in line 567 referring back to the bug for the reasons of not releasing the GIL for functions likely called on the exit path? 

There are many other places that may release GIL and may sit on the exit path. We recently encountered a deadlock that also tracks back to the release of GIL near exit, but the origin is from the dealloc of an extension type. I also know that freeing Tensor object releases GIL too, and not releasing GIL from that path could potentially become a performance regression.

My understanding is that there are changes in Python upstream (but I don't immediately recall the version 3.10?) that makes GIL use during exit safer in general. We haven't tried if it can solve the deadlock issue. I wonder if it is an option for you to try with a newer CPython version?

Flamefire commented:
@rainwoodman 
> Could you add a comment in line 567 referring back to the bug for the reasons of not releasing the GIL for functions likely called on the exit path?

Sorry for the delay, I added the requested comment and rebased onto current master.

gbaned commented:
Hi @rainwoodman Can you please review this PR ? Thank you!

rainwoodman commented:
Thanks for the update. Sorry for the late approval. This almost missed my inboxes.
